File: typescript/src/financialAction.ts
/**
 * Financial Action Model — canonical types for financial execution authority.
 *
 * Defines the core vocabulary for all financial actions governed by
 * AtlaSent's economic governance layer. Every consequential financial
 * operation must be classified here before execution authorization.
 *
 * Wire-stable as `financial_action.v1`.
 */

/** ISO 4217 currency code. */
export type CurrencyCode =
  | "USD" | "EUR" | "GBP" | "JPY" | "CAD" | "AUD" | "CHF"
  | "CNY" | "SEK" | "NZD" | string;

/**
 * Risk tier for a financial action.
 *
 * - low:      < $1,000 — single-approver sufficient
 * - medium:   $1,000–$50,000 — dual-control typically required
 * - high:     $50,000–$1,000,000 — CFO-level required
 * - critical: > $1,000,000 or irreversible — board-level or emergency protocol
 */
export type FinancialRiskTier = "low" | "medium" | "high" | "critical";

/** How accountability is distributed across parties for a financial action. */
export type LiabilityClassification =
  | "individual"          // Single party bears full liability
  | "shared"              // Multiple parties share liability proportionally
  | "delegated"           // Liability flows to the delegate, not delegator
  | "supervisory"         // Supervisors bear liability for team actions
  | "emergency_override"; // Special liability regime for emergency bypasses

/** Canonical set of financial action types. */
export type FinancialActionType =
  | "refund"
  | "payment_release"
  | "invoice_approval"
  | "payroll_execution"
  | "procurement_approval"
  | "wire_transfer"
  | "trading_execution"
  | "budget_override"
  | "spending_authorization"
  | "vendor_payment"
  | "subscription_cancellation"
  | "credit_issuance"
  | "fee_waiver"
  | "chargeback"
  | "contract_commitment"
  | string;

/**
 * Canonical definition of a financial action class.
 *
 * Stored in `financial_action_classes`. Drives quorum policy,
 * liability classification, and risk tier assignment.
 */
export interface FinancialActionClass {
  /** Stable identifier (e.g. "wire_transfer.domestic"). */
  readonly action_class_id: string;
  readonly name: string;
  readonly action_type: FinancialActionType;
  readonly risk_tier: FinancialRiskTier;
  /** Minimum number of approvals before execution is permitted. */
  readonly required_approvals: number;
  readonly liability_classification: LiabilityClassification;
  /** Whether this action is reversible post-execution. */
  readonly reversible: boolean;
  /** Maximum allowed execution value for autonomous agents (null = no ceiling). */
  readonly autonomous_ceiling: number | null;
  readonly ceiling_currency: CurrencyCode | null;
  readonly created_at: string;
  readonly description?: string;
}

/** Status of a financial execution record. */
export type FinancialExecutionStatus =
  | "pending_approval"
  | "approved"
  | "executing"
  | "completed"
  | "failed"
  | "reversed"
  | "disputed"
  | "frozen";

/**
 * Immutable record of a financial action execution.
 *
 * Written at authorization time. Stored in `financial_execution_records`.
 */
export interface FinancialExecutionRecord {
  readonly execution_id: string;
  readonly action_class_id: string;
  readonly org_id: string;
  readonly action_value: number;
  readonly currency: CurrencyCode;
  readonly risk_tier: FinancialRiskTier;
  readonly liability_classification: LiabilityClassification;
  readonly initiator_id: string;
  readonly executor_id: string;
  /** Ordered list of approver IDs. */
  readonly approver_ids: readonly string[];
  /** AtlaSent permit IDs — one per approval stage. */
  readonly permit_ids: readonly string[];
  readonly override_applied: boolean;
  readonly override_id: string | null;
  readonly status: FinancialExecutionStatus;
  readonly authorized_at: string;
  readonly executed_at: string | null;
  /** Hash-chained audit trail entry. */
  readonly audit_hash: string;
  readonly context: Record<string, unknown>;
}

/** Threshold configuration for risk-tier escalation. */
export interface RiskTierThreshold {
  readonly tier: FinancialRiskTier;
  /** Inclusive lower bound in the reference currency. */
  readonly lower_bound: number;
  /** Exclusive upper bound; null means unbounded. */
  readonly upper_bound: number | null;
  readonly reference_currency: CurrencyCode;
}

/** Default risk tier thresholds (USD-denominated). */
export const DEFAULT_RISK_TIER_THRESHOLDS: readonly RiskTierThreshold[] = [
  { tier: "low",      lower_bound: 0,           upper_bound: 1_000,       reference_currency: "USD" },
  { tier: "medium",   lower_bound: 1_000,       upper_bound: 50_000,      reference_currency: "USD" },
  { tier: "high",     lower_bound: 50_000,      upper_bound: 1_000_000,   reference_currency: "USD" },
  { tier: "critical", lower_bound: 1_000_000,   upper_bound: null,        reference_currency: "USD" },
] as const;

/**
 * Classify a financial action's risk tier based on its value.
 */
export function classifyRiskTier(
  value: number,
  thresholds: readonly RiskTierThreshold[] = DEFAULT_RISK_TIER_THRESHOLDS,
): FinancialRiskTier {
  for (const t of thresholds) {
    if (value >= t.lower_bound && (t.upper_bound === null || value < t.upper_bound)) {
      return t.tier;
    }
  }
  return "critical";
}

/**
 * Return true when the action value is within the autonomous execution ceiling.
 * A null ceiling means no ceiling is configured (always within bounds).
 */
export function withinAutonomousCeiling(
  actionValue: number,
  ceiling: number | null,
): boolean {
  if (ceiling === null) return true;
  return actionValue <= ceiling;
}

File: python/atlasent/governance/financial_action.py
"""Financial Action Model — canonical types for financial execution authority.

Mirrors ``atlasent-sdk/typescript/src/financialAction.ts``. Wire-stable as
``financial_action.v1``. Schema lives in ``contract/schemas/`` (pending) and
is enforced by migration ``001_financial_action_model.sql``.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import Any, Literal

# ─── Type aliases ────────────────────────────────────────────────────

# Open string union: ISO 4217 well-known codes plus any other 3-letter code.
CurrencyCode = str

FinancialRiskTier = Literal["low", "medium", "high", "critical"]

LiabilityClassification = Literal[
    "individual",
    "shared",
    "delegated",
    "supervisory",
    "emergency_override",
]

FinancialActionType = (
    str  # canonical set documented in TS; open-ended for extensibility
)

FinancialExecutionStatus = Literal[
    "pending_approval",
    "approved",
    "executing",
    "completed",
    "failed",
    "reversed",
    "disputed",
    "frozen",
]


@dataclass(frozen=True)
class FinancialActionClass:
    """Canonical per-org definition of a financial action class.

    Stored in ``financial_action_classes``. Drives quorum policy,
    liability classification, and risk-tier assignment.
    """

    action_class_id: str
    name: str
    action_type: FinancialActionType
    risk_tier: FinancialRiskTier
    required_approvals: int
    liability_classification: LiabilityClassification
    reversible: bool
    autonomous_ceiling: float | None
    ceiling_currency: CurrencyCode | None
    created_at: str
    description: str | None = None


@dataclass(frozen=True)
class FinancialExecutionRecord:
    """Immutable record of a financial action execution.

    Written at authorization time. Stored append-only in
    ``financial_execution_records``.
    """

    execution_id: str
    action_class_id: str
    org_id: str
    action_value: float
    currency: CurrencyCode
    risk_tier: FinancialRiskTier
    liability_classification: LiabilityClassification
    initiator_id: str
    executor_id: str
    approver_ids: Sequence[str]
    permit_ids: Sequence[str]
    override_applied: bool
    override_id: str | None
    status: FinancialExecutionStatus
    authorized_at: str
    executed_at: str | None
    audit_hash: str
    context: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class RiskTierThreshold:
    """Threshold configuration for risk-tier escalation."""

    tier: FinancialRiskTier
    lower_bound: float
    upper_bound: float | None
    reference_currency: CurrencyCode = "USD"


# Default risk tier thresholds (USD-denominated). Mirrors the TS
# ``DEFAULT_RISK_TIER_THRESHOLDS`` constant exactly.
DEFAULT_RISK_TIER_THRESHOLDS: tuple[RiskTierThreshold, ...] = (
    RiskTierThreshold(
        tier="low", lower_bound=0.0, upper_bound=1_000.0, reference_currency="USD"
    ),
    RiskTierThreshold(
        tier="medium",
        lower_bound=1_000.0,
        upper_bound=50_000.0,
        reference_currency="USD",
    ),
    RiskTierThreshold(
        tier="high",
        lower_bound=50_000.0,
        upper_bound=1_000_000.0,
        reference_currency="USD",
    ),
    RiskTierThreshold(
        tier="critical",
        lower_bound=1_000_000.0,
        upper_bound=None,
        reference_currency="USD",
    ),
)


def classify_risk_tier(
    value: float,
    thresholds: Sequence[RiskTierThreshold] = DEFAULT_RISK_TIER_THRESHOLDS,
) -> FinancialRiskTier:
    """Classify a financial action's risk tier based on its value.

    Iteration order matches the TS implementation: first matching threshold wins.
    Values that fall below all thresholds map to ``"low"`` (when default thresholds
    are used and value >= 0); values above the highest upper_bound fall through to
    the implicit ``"critical"`` final return.
    """
    for t in thresholds:
        if value >= t.lower_bound and (t.upper_bound is None or value < t.upper_bound):
            return t.tier
    return "critical"


def within_autonomous_ceiling(action_value: float, ceiling: float | None) -> bool:
    """Return True iff action_value is within the autonomous-execution ceiling.

    A ``None`` ceiling means no ceiling configured (always within bounds).
    """
    if ceiling is None:
        return True
    return action_value <= ceiling


__all__ = [
    "DEFAULT_RISK_TIER_THRESHOLDS",
    "CurrencyCode",
    "FinancialActionClass",
    "FinancialActionType",
    "FinancialExecutionRecord",
    "FinancialExecutionStatus",
    "FinancialRiskTier",
    "LiabilityClassification",
    "RiskTierThreshold",
    "classify_risk_tier",
    "within_autonomous_ceiling",
]
